File: php/kruskal.php
<?php

class DisjointSet {
    private $parent;
    private $rank;

    public function __construct($n) {
        $this->parent = range(0, $n - 1);
        $this->rank = array_fill(0, $n, 0);
    }

    public function find($x) {
        if ($this->parent[$x] !== $x) {
            $this->parent[$x] = $this->find($this->parent[$x]);
        }
        return $this->parent[$x];
    }

    public function union($x, $y) {
        $rootX = $this->find($x);
        $rootY = $this->find($y);

        if ($rootX !== $rootY) {
            if ($this->rank[$rootX] > $this->rank[$rootY]) {
                $this->parent[$rootY] = $rootX;
            } elseif ($this->rank[$rootX] < $this->rank[$rootY]) {
                $this->parent[$rootX] = $rootY;
            } else {
                $this->parent[$rootY] = $rootX;
                $this->rank[$rootX]++;
            }
        }
    }
}

class Graph {
    private $V;
    private $edges;

    public function __construct($vertices) {
        $this->V = $vertices;
        $this->edges = [];
    }

    public function addEdge($u, $v, $w) {
        $this->edges[] = [$u, $v, $w];
    }

    public function kruskalMST() {
        usort($this->edges, function($a, $b) {
            return $a[2] - $b[2];
        });

        $disjointSet = new DisjointSet($this->V);
        $result = [];

        foreach ($this->edges as [$u, $v, $w]) {
            $setU = $disjointSet->find($u);
            $setV = $disjointSet->find($v);

            if ($setU !== $setV) {
                $result[] = [$u, $v, $w];
                $disjointSet->union($setU, $setV);
            }
        }

        return $result;
    }
}

// Example usage:
$graph = new Graph(4);
$graph->addEdge(0, 1, 10);
$graph->addEdge(0, 2, 6);
$graph->addEdge(0, 3, 5);
$graph->addEdge(1, 3, 15);
$graph->addEdge(2, 3, 4);

$mst = $graph->kruskalMST();
echo "Edges in the Minimum Spanning Tree:\n";
foreach ($mst as [$u, $v, $w]) {
    echo "($u, $v) -> $w\n";
}

File: python/kruskal.py
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x != root_y:
            if self.rank[root_x] > self.rank[root_y]:
                self.parent[root_y] = root_x
            elif self.rank[root_x] < self.rank[root_y]:
                self.parent[root_x] = root_y
            else:
                self.parent[root_y] = root_x
                self.rank[root_x] += 1

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.edges = []

    def add_edge(self, u, v, w):
        self.edges.append((u, v, w))

    def kruskal_mst(self):
        self.edges.sort(key=lambda x: x[2])
        disjoint_set = DisjointSet(self.V)
        result = []

        for u, v, w in self.edges:
            set_u = disjoint_set.find(u)
            set_v = disjoint_set.find(v)

            if set_u != set_v:
                result.append((u, v, w))
                disjoint_set.union(set_u, set_v)

        return result

# Example usage:
graph = Graph(4)
graph.add_edge(0, 1, 10)
graph.add_edge(0, 2, 6)
graph.add_edge(0, 3, 5)
graph.add_edge(1, 3, 15)
graph.add_edge(2, 3, 4)

mst = graph.kruskal_mst()
print("Edges in the Minimum Spanning Tree:")
for u, v, w in mst:
    print(f"({u}, {v}) -> {w}")
